This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

In [1]:
# get api key from your .env file
import os
from dotenv import load_dotenv

load_dotenv()
API_KEY = os.getenv('NASDAQ_API_KEY')


print(API_KEY)

9_fVkE8xTZRMwTqYLB9A


Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.


You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

In [2]:
%pip install requests

Note: you may need to restart the kernel to use updated packages.


Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [3]:
# First, import the relevant modules
import requests, collections, json
from pprint import pprint as pp

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [4]:
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned

# put your params
my_params = {'limit': '1'}

r = requests.get("https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X.json?api_key=9_fVkE8xTZRMwTqYLB9A",params=my_params)

print(type(r.text))

<class 'str'>


In [5]:
# converts the class str in var r and converts it to class dic 
data = json.loads(r.text)
print(type(data))

# formats betters, but turns dic into string
data = json.dumps(data, indent=4, sort_keys=True)
print(type(data))

print(data)

<class 'dict'>
<class 'str'>
{
    "dataset": {
        "collapse": null,
        "column_index": null,
        "column_names": [
            "Date",
            "Open",
            "High",
            "Low",
            "Close",
            "Change",
            "Traded Volume",
            "Turnover",
            "Last Price of the Day",
            "Daily Traded Units",
            "Daily Turnover"
        ],
        "data": [
            [
                "2020-12-01",
                112.2,
                112.2,
                111.5,
                112.0,
                null,
                51.0,
                5703.0,
                null,
                null,
                null
            ]
        ],
        "database_code": "FSE",
        "database_id": 6129,
        "dataset_code": "AFX_X",
        "description": "Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704",
        "end_

In [6]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).

In [7]:
p = {'start_date':'2017-01-01','end_date':'2017-12-31'}
source = requests.get("https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X.json?api_key=9_fVkE8xTZRMwTqYLB9A",
                 params=p)

# converts the class str in var r and converts it python object
data = json.loads(source.text)

print(json.dumps(data, indent=3, sort_keys=True))


{
   "dataset": {
      "collapse": null,
      "column_index": null,
      "column_names": [
         "Date",
         "Open",
         "High",
         "Low",
         "Close",
         "Change",
         "Traded Volume",
         "Turnover",
         "Last Price of the Day",
         "Daily Traded Units",
         "Daily Turnover"
      ],
      "data": [
         [
            "2017-12-29",
            51.76,
            51.94,
            51.45,
            51.76,
            null,
            34640.0,
            1792304.0,
            null,
            null,
            null
         ],
         [
            "2017-12-28",
            51.65,
            51.82,
            51.43,
            51.6,
            null,
            40660.0,
            2099024.0,
            null,
            null,
            null
         ],
         [
            "2017-12-27",
            51.45,
            51.89,
            50.76,
            51.82,
            null,
            57452.0,
        

limit_output extension: Maximum message size of 10000 exceeded with 62117 characters

2. Convert the returned JSON object into a Python dictionary.

In [8]:
print(type(data))

for item in data['dataset']:
    print(item)

<class 'dict'>
id
dataset_code
database_code
name
description
refreshed_at
newest_available_date
oldest_available_date
column_names
frequency
type
premium
limit
transform
column_index
start_date
end_date
data
collapse
order
database_id


In [9]:
data = data['dataset']

In [10]:
data['end_date']

'2017-12-31'

In [11]:
pp(data)

{'collapse': None,
 'column_index': None,
 'column_names': ['Date',
                  'Open',
                  'High',
                  'Low',
                  'Close',
                  'Change',
                  'Traded Volume',
                  'Turnover',
                  'Last Price of the Day',
                  'Daily Traded Units',
                  'Daily Turnover'],
 'data': [['2017-12-29',
           51.76,
           51.94,
           51.45,
           51.76,
           None,
           34640.0,
           1792304.0,
           None,
           None,
           None],
          ['2017-12-28',
           51.65,
           51.82,
           51.43,
           51.6,
           None,
           40660.0,
           2099024.0,
           None,
           None,
           None],
          ['2017-12-27',
           51.45,
           51.89,
           50.76,
           51.82,
           None,
           57452.0,
           2957018.0,
           None,
           None,
          

limit_output extension: Maximum message size of 10000 exceeded with 22194 characters

In [12]:
type(data)

dict

In [13]:
data.keys()

dict_keys(['id', 'dataset_code', 'database_code', 'name', 'description', 'refreshed_at', 'newest_available_date', 'oldest_available_date', 'column_names', 'frequency', 'type', 'premium', 'limit', 'transform', 'column_index', 'start_date', 'end_date', 'data', 'collapse', 'order', 'database_id'])

In [14]:
data['column_names']

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

In [15]:
len(data['data'])

255

In [16]:
# we want high, low, close, traded volumne 
data['column_names']

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

3. Calculate what the highest and lowest **opening prices** were for the stock in this period.

In [17]:
data['data'][0]

['2017-12-29',
 51.76,
 51.94,
 51.45,
 51.76,
 None,
 34640.0,
 1792304.0,
 None,
 None,
 None]

In [35]:
open_price = list()
high = list()
low = list()
closing_price = list()
traded_vol = list()

In [36]:

# open_price = [[data['data'][data][1]] for date in range(len(data['data'])) if 
#               [data['data'][data][1]] is None]


for date in range(len(data['data'])):
    open_price.append(data['data'][date][1])


#     open_price = [data['data'][data][1]] if data['data'][data][1] != None]
    high.append(data['data'][date][2])
    low.append(data['data'][date][3])
    closing_price.append(data['data'][date][4])
    traded_vol.append(data['data'][date][6])
#     print(open_price, high, low, closing_price, traded_vol)


In [37]:
# removes Nonetype in open_price
open_price = [price for price in open_price if price is not None]
print(open_price)

[51.76, 51.65, 51.45, 51.05, 51.16, 51.88, 52.73, 52.37, 52.7, 53.11, 52.64, 52.29, 52.28, 51.5, 50.89, 50.8, 51.21, 49.5, 49.52, 48.64, 49.64, 49.09, 49.13, 49.11, 48.8, 48.4, 47.25, 46.57, 47.03, 47.09, 47.98, 48.4, 48.38, 47.3, 47.65, 46.42, 46.16, 45.81, 45.0, 45.88, 46.29, 46.53, 45.48, 45.2, 45.01, 45.16, 44.9, 45.08, 45.72, 46.01, 45.8, 45.61, 45.5, 45.58, 45.97, 45.64, 46.2, 46.19, 46.01, 45.36, 44.51, 43.58, 42.0, 42.35, 42.3, 42.3, 41.48, 42.29, 42.54, 42.65, 42.5, 42.29, 42.35, 42.49, 43.21, 42.81, 42.7, 43.0, 42.66, 43.0, 42.38, 42.16, 42.0, 42.0, 41.71, 42.11, 42.64, 42.72, 42.82, 42.46, 42.42, 42.28, 41.88, 42.4, 42.53, 42.12, 41.3, 41.73, 43.5, 44.9, 45.85, 45.13, 45.34, 45.25, 45.24, 44.94, 45.26, 45.16, 44.91, 44.7, 45.31, 45.57, 45.74, 45.06, 45.5, 45.6, 45.07, 44.67, 44.29, 44.94, 44.64, 44.79, 45.5, 44.67, 45.83, 45.29, 45.01, 45.73, 46.68, 47.23, 46.95, 47.29, 47.03, 47.46, 46.48, 46.9, 45.66, 46.34, 46.52, 46.5, 47.31, 46.77, 47.8, 47.01, 47.12, 46.8, 46.12, 45.22

In [38]:
# highest open_price
max(open_price)

53.11

In [39]:
min(open_price)

34.0

The highest opening price is **53.11** and the lowest opening price is **34.00**.

#### Question 4: What was the largest change in any one day (based on High and Low price)?

```2.81```

In [40]:
# holds high price - low price 
day_change = list()

for h, l in zip(high,low):
    day_change.append(h-l)
    
print(round(sorted(day_change)[-1],2))

2.81


### 5. What was the largest change between any two days (based on Closing Price)?

2.56

In [41]:
two_days_difference = list() 

In [42]:
range(len(closing_price))


range(0, 255)

In [43]:
for x in range(3):
    print(x)

0
1
2


In [44]:
two_days = [closing_price[x+1]-closing_price[x] 
            for x in range(len(closing_price)-1)]

In [45]:
round(max(two_days),2)

2.56

#### 6. What was the average daily trading volume during this year

89,123.34 shares

In [46]:
from statistics import mean

In [49]:
round(mean(traded_vol),2)

89124.34

7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

For the year 2017, AFX_X has 255 entries. 

```76,286.00 shares```


In [54]:
def median(traded_vol):
    traded_vol.sort()
    
    if len(traded_vol) % 2 != 0:
        median = int((len(traded_vol)-1)/2)
        return traded_vol[median]
    
    elif len(traded_vol) % 2 ==0:
        median_idx_1 = int(len(traded_vol) / 2)
        median_idx_2 = int(len(traded_vol) / 2) -1 
        return int(mean(traded_vol[median_idx_1], 
                        traded_vol[median_idx_2]))
    
    


In [55]:
median(traded_vol)

76286.0